### PR TITLE
support terraform 0.12

### PR DESCRIPTION
**Pull request by hyww**
#### Changes
+ use `terraform 0.12upgrade` command of terraform 0.12.3 to upgrade syntax of this module
+ fix type constraint of variable `backends` (the above command set it to `map(string)`, which is not correct)

#### Reference
+ [Upgrading to Terraform 0.12 - Terraform by HashiCorp](https://www.terraform.io/upgrade-guides/0-12.html)

Fixes #49

### CODE REVIEW

**Comment by morgante** [comment]
The examples also need to be updated to support Terraform v0.12.

@ivankorn Please go ahead and start a new PR which incorporates this one but also updates examples.

**Comment by ivankorn** [comment]
> @ivankorn Please go ahead and start a new PR which incorporates this one but also updates examples.

sure, I'll start it as #51 on Monday then

File: main.tf [hunk 1]
```diff
@@ -15,99 +15,149 @@
...
 
 # HTTP proxy when ssl is false
 resource "google_compute_target_http_proxy" "default" {
-  project = "${var.project}"
-  count   = "${var.http_forward ? 1 : 0}"
+  project = var.project
+  count   = var.http_forward ? 1 : 0
   name    = "${var.name}-http-proxy"
-  url_map = "${element(compact(concat(list(var.url_map), google_compute_url_map.default.*.self_link)), 0)}"
+  url_map = element(
```
Comment (itscaro) [inline main.tf]: maybe `compact(concat(list(var.url_map), google_compute_url_map.default.*.self_link))[0]`

File: main.tf [hunk 2]
```diff
@@ -15,99 +15,149 @@
...
 resource "google_compute_target_https_proxy" "default" {
-  project          = "${var.project}"
-  count            = "${var.ssl ? 1 : 0}"
-  name             = "${var.name}-https-proxy"
-  url_map          = "${element(compact(concat(list(var.url_map), google_compute_url_map.default.*.self_link)), 0)}"
-  ssl_certificates = ["${compact(concat(var.ssl_certificates, google_compute_ssl_certificate.default.*.self_link))}"]
+  project = var.project
+  count   = var.ssl ? 1 : 0
+  name    = "${var.name}-https-proxy"
+  url_map = element(
```
Comment (itscaro) [inline main.tf]: maybe `compact(concat(list(var.url_map), google_compute_url_map.default.*.self_link))[0]`

File: variables.tf [hunk 3]
```diff
@@ -14,104 +14,105 @@
...
   description = "List of target tags for health check firewall rule."
-  type        = "list"
+  type        = list(string)
 }
 
-variable backends {
+variable "backends" {
   description = "Map backend indices to list of backend maps."
-  type        = "map"
+  type        = map
```
Comment (itscaro) [inline variables.tf]: IMO, this should become a list as backend_params is a list, but this is not what this PR is about :)

**Pull request closed without merging**